### DISCUSSION TITLE
The `<:` assignment operator, and the `:>` flow operator

### DISCUSSION THREAD

**Discussion by getify** [getify/foi-lang#6] [Ideas]
I really liked the semantic showing the direction-of-information in an assignment that `<:` gives. Moreover, `:>` is currently the flow operator (compose-left-to-right). Initially, I thought that was kind of a cool symmetry with the assignment operator.

But the more I interact with `<  >` Records/Tuples, the more I find these operators using the `<` and `>` characters perhaps being more confusing. Perhaps the intended semantics are too subtle/clever to matter, and the similarity will just be confusing in the long run.

So...

1. What would be a better assignment operator?

    I am pretty firmly against using `=` in any way for that purpose, as I've always disliked the conflation with assignment and equality-comparison. `?=` and `!=` are the equality/inequality operators already, and I don't see that changing.

    I kind of think that `:` should be part of the assignment operator, since `:` is used for initial `def` definitions, as well as field name definitions in Records.

    What about `::`? Is that too weird?

    As a FP language, **Foi** needs to straddle a fine line here. Re-assignment must be possible, but it should stand out and be pretty obvious that it's happening.

2. What would be a better flow (flow-right) operator? Or is `:>` fine as-is, if we already change the assignment operator?

    Originally, I had `>>` but then realized that would have potential grammar ambiguity with the closing tags of nested Records/Tuples.

    Also, the pipeline operator is `#>` and I really like that because then the `#` is available as the topic token in the pipeline.

**Comment by jschatz1** [comment]
`::` I associate with Rust and others like it. Even though you don't like or agree with `=`, it's super standard and I feel like you could move on to more important issues and have a wider adoption if you just use `=`. It's assignment functionality is pretty well ingrained. Anything else might be hard to get to stick. 
If you are going for something else I always liked `<-`. 

`a <-5`, in that you are storing `5` in memory location you are calling `a`. I think of it in reverse to the average. e.g. you are placing the value in memory, not that you are assigning the variable the value. 

If you are going the other direction you could do: `a -> 5`. 

I'm sure there's more you can do.

**Reply by getify** [reply]
What do you think about the pascal `:=` option?

**Reply by getify** [reply]
I posted an update just now... it switched assignment to `:=` and it switched flow/compose operators to `+>` and `<+` respectively. We'll see how this feels.
